### PR TITLE
Add utility functions for reading testing env vars

### PR DESCRIPTION
**Pull request by aznashwan**
Depends on https://github.com/canonical/k8s-workflows/pull/15

### CODE REVIEW

File: k8s_test_harness/util/docker_util.py [hunk 1]
```diff
@@ -4,11 +4,35 @@
...
-            ["docker", "run", "--rm", image, "ls", "-l", path],
-            check=True,
-        )
+    """Ensures the given container image contains the provided paths
+    within it by attempting to run the image and 'ls' all of them.
+    """
+    base_cmd = ["docker", "run", "--rm", image, "ls", "-l"]
+    base_cmd.extend(paths)
+
+    subprocess.run(base_cmd, check=True)
```
Comment (bschimke95) [inline k8s_test_harness/util/docker_util.py:17]: nit: you can use https://github.com/canonical/k8s-test-harness/blob/d503366f0b481183567de806175ac8d6fee547dc/k8s_test_harness/util/exec_util.py#L25 for this.

Comment (bschimke95) [reply k8s_test_harness/util/docker_util.py:17]: and below.

File: k8s_test_harness/util/env_util.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    return val
+
+
+@dataclass
+class RockMetaInfo:
+    """Class representing metadata for a built ROCK as produced by the
+    canonical/k8s-workflows/build_rocks.yaml workflow:
+
+    # TODO(aznashwan): pin exact link upon first stable release of workflows:
+    https://github.com/canonical/k8s-workflows/blob/main/.github/workflows/build_rocks.yaml#L125-L133
```
Comment (bschimke95) [inline k8s_test_harness/util/env_util.py:41]: Just a reminder comment to update before merging this

File: k8s_test_harness/util/env_util.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            if init_kwargs[kwarg] == None]
+        if unset_keys:
+            raise ValueError(
+                f"Missing ROCK build meta info fields {unset_keys} from "
+                f"the provided dict: {dict_val}")
+
+        return cls(**init_kwargs)  # pyright: ignore
+
+    @classmethod
+    def from_json_string(cls, json_str: str) -> 'RockMetaInfo':
```
Comment (bschimke95) [inline k8s_test_harness/util/env_util.py:90]: nit: you can use `typing.Self` here.

File: k8s_test_harness/util/env_util.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    return [RockMetaInfo.from_dict(d) for d in rock_metas_list]
+
+
+def get_builds_meta_info_for_rock(
+        rock_name: str,
+        rock_metadata_env_var: str=DEFAULT_BUILT_ROCKS_METADATA_ENV_VAR
+) -> List[RockMetaInfo]:
+    """Returns a list of all build meta info sets for the ROCK with
+    the given name (should be one entry per version and architecture
+    the ROCK was built for) from the ;environment variable produced
```
Comment (bschimke95) [inline k8s_test_harness/util/env_util.py]: nit: extra `;`

File: k8s_test_harness/util/env_util.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    @property
+    def rockfile_path(self) -> str:
+        """Returns the path to the rockfile.yaml file for this image
+        based on the `path` property (RockMetaInfo.rock_dir) of the
+        ROCK's meta info.
+
+        Note that this can be a relative path with respect to the root
+        of the repository the `build_rocks.yaml` workflow had run on.
+        """
+        return os.path.join(self.rock_dir, "rockfile.yaml")
```
Comment (berkayoz) [inline k8s_test_harness/util/env_util.py]:
```suggestion
        return os.path.join(self.rock_dir, "rockcraft.yaml")
```
Isn't this supposed to be `rockcraft.yaml`?

**Review by berkayoz — COMMENTED** [review]
Looks good overall, left a comment

File: k8s_test_harness/util/env_util.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+    return [r for r in all_metas if r.name == rock_name]
+
+
+def get_build_meta_info_for_rock_version(
+        rock_name: str,
+        rock_version: str,
+        rock_arch: str,
+        rock_metadata_env_var: str=DEFAULT_BUILT_ROCKS_METADATA_ENV_VAR
+) -> RockMetaInfo:
```
Comment (bschimke95) [inline k8s_test_harness/util/env_util.py:164]: just a comment: this basically rebuilds the whole `RockMetaInfo` list from the environment every time you call this function. Maybe returning a `RockMetaInfos` data object would make sense and this object has a `get_info_for_rock_version`. 

Anyway, this is not performance critical code - just leave it like this.

**Review by bschimke95 — COMMENTED** [review]
Minor nits (and a comment) then we are good to merge (after the workflow is merged).

Great work!

File: k8s_test_harness/util/env_util.py (hunk 3 above)
Comment (aznashwan) [reply k8s_test_harness/util/env_util.py:90]: That was [what I did initially](https://github.com/canonical/k8s-test-harness/pull/5/commits/9e524854e5927536b80231ee02100951d0efce99) but the GH runner is using Python 3.8 where it does not exist and I had to switch to using a plain string for the code to actually run. (in terms of type hinting tooling the effect should be the same)

File: k8s_test_harness/util/env_util.py (hunk 2 above)
Comment (aznashwan) [reply k8s_test_harness/util/env_util.py:41]: This requires a tag/release on the workflows (which is why the URL is currently linking to the `main` branch).

Without it, I could either link the current HEAD of the repo or remove the comment entirely.

File: k8s_test_harness/util/env_util.py (hunk 5 above)
Comment (aznashwan) [reply k8s_test_harness/util/env_util.py]: Great catch, brain fart, updated the function name for clarity as well!

File: k8s_test_harness/util/docker_util.py (hunk 1 above)
Comment (aznashwan) [reply k8s_test_harness/util/docker_util.py:17]: Sadly there's something weird with how pytest loads modules going on and it [leads to partially initalized import issues](https://github.com/aznashwan/harbor-rocks/actions/runs/9989824882/job/27609690604#step:7:53).

In the interest of time, I have converted it back to using `subprocess` directly. (the added logging that `exec_util.run()` offers won't be printed on successful tests anyway, and the whole command line argument list will be reported by `subprocess` on errors anyway, so there's no real notable disadvantage without it anyway)

File: k8s_test_harness/util/env_util.py (hunk 6 above)
Comment (aznashwan) [reply k8s_test_harness/util/env_util.py:164]: Agreed, but most sanity tests will only call this once anyways, and I can add it later if I feel the integration tests could really use it.

**Review by bschimke95 — APPROVED** [review]
LGTM

**Pull request merged**